lms/djangoapps/class_dashboard/dashboard_data.py

```diff
@@ -529,29 +558,38 @@ def get_students_opened_subsection(request, csv=False):
     student names, usernames for CSV download.
     """
     module_state_key = Location.from_deprecated_string(request.GET.get('module_id'))
+    course_id = request.GET.get('course_id')
     csv = request.GET.get('csv')
 
     # Query for "opened a subsection" students
     students = models.StudentModule.objects.select_related('student').filter(
         module_state_key__exact=module_state_key,
         module_type__exact='sequential',
-    ).values('student__username', 'student__profile__name').order_by('student__profile__name')
+    ).values('student__id', 'student__username', 'student__profile__name').order_by('student__profile__name')
+
+    course_key = locator.CourseLocator.from_string(course_id)
+    non_student_id_list = get_non_student_list(course_key)  # course_key is not a string
+    num_non_student = len(non_student_id_list)
 
     results = []
+    non_student_skipped = 0
     if not csv:
         # Restrict screen list length
-        # Adding 1 so can tell if list is larger than MAX_SCREEN_LIST_LENGTH
+        # Adding 1 + num_non_student so can tell if list is larger than MAX_SCREEN_LIST_LENGTH. non-students will be skipped.
         # without doing another select.
-        for student in students[0:MAX_SCREEN_LIST_LENGTH + 1]:
-            results.append({
-                'name': student['student__profile__name'],
-                'username': student['student__username'],
-            })
-
+        for student in students[0:MAX_SCREEN_LIST_LENGTH + 1 + num_non_student]:
```

Comment:
@lydiali 
There's several ways of doing this.
1. Use select_related to do multiple joins from courseware_studentmodule (csm) to user to courseaccessrole. I haven't actually tried this but it looks possible just by eyeballing the relations in the models.
2. In the code, above (e.g. line 71), you are splitting it into 2 queries with the csm containing a
   .exclude(student_id__in=non_student_list)
clause.
3. Subtract the subset from the csm list using Python lists. This is how you currently have it.

I believe (1) is a really bad idea. csm is the most poorly designed table in EdX. I just tried to find out how many rows it has in production but the query timed out after 10 minutes so we'll never know :-1:  So, doing anything that makes the query longer running doesn't seem good. Also, for over 99% of the rows returned the join is superfluous; the student is not Staff or Instructor and so it seems an inefficient join.

I like (2). There are no extra joins, just an extra "where" clause and is consistent with what you're doing in the code above. Since the non-student list will be very small it shouldn't make the query any worse.

(3) Seems as though it should be equivalent to (2) in many respects.

I'm running with production data in my dev system so I could do a benchmark if you wanted. Otherwise, I'm thinking (2) would be best unless I've missed something.

Looking to the future, I believe this bit of code may get re-written when the Insights api supports student level, non aggregated, data. @dylanrhodes may be able to confirm that or let us know the timescales of when that is likely.